### ISSUE TITLE
`parserOptions.project` doesnt handle project references well

### ISSUE THREAD

**Issue by milesj** [typescript-eslint/typescript-eslint#856]
**What code were you trying to parse?**

Not parsing code exactly, but I'm attempting to upgrade to v2, which hard fails if `project` doesn't include files correctly. This doesn't seem to handle project references correctly.

Given this root `tsconfig.json`:

```
{
  "files": [],
  "references": [
    {
      "path": "packages/config-danger"
    }
  ],
  "extends": "./tsconfig.options.json"
}
```

And this `config-danger/tsconfig.json`:

```
{
  "compilerOptions": {
    "declarationDir": "lib",
    "outDir": "lib",
    "rootDir": "src"
  },
  "exclude": [
    "lib",
    "tests"
  ],
  "extends": "../../tsconfig.options.json",
  "include": [
    "src/**/*",
    "types/**/*",
    "../../types/**/*"
  ],
  "references": []
}
```

**What did you expect to happen?**

Files within a project reference `include` worked correctly.

**What actually happened?**

The parser/plugin fails with a ton of the following errors:

```
/root/packages/config-danger/src/code.ts
  0:0  error  Parsing error: If "parserOptions.project" has been set for @typescript-eslint/parser, /root/packages/config-danger/src/code.ts must be included in at least one of the projects provided
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.0.0` |
| `TypeScript`                | `3.5.2` |
| `ESLint`                    | `6.1.0` |
| `node`                      | `12.8.0` |
| `npm`                       | `6.1.0.3` |

**Comment by ruscon** [comment]
I have similar issue too.

**Comment by ruscon** [comment]
I have found a reason.

```json5
"allowJs": true,
```
If not specified in the `tsconfig.json` or in the `tsconfig.eslint.json`, then js files do not want to be processed.

**Comment by bradzacher** [comment]
Hmm, I'm not sure what the typescript compiler actually does to handle project references. I haven't actually used them yet (I've been meaning to migrate this monorepo to them, but I keep forgetting).

Perhaps @uniqueiniquity has more of an idea than I do.
Is TS supposed to "include" all of the files in the referenced projects, or are they consumed via a different mechanism?

@milesj - could you please try using the array form of `project` to specify all of your tsconfigs?
```json
{
  "parserOptions": {
    "project": [
      "tsconfig.json",
      "packages/config-danger/tsconfig.json"
    ]
  }
}
```

**Comment by 19majkel94** [comment]
I can confirm that the array of projects work but it's really making its hard to maintain:

```js
    project: [
      path.resolve(__dirname, "./tsconfig.json"),
      path.resolve(__dirname, "./packages/core/tsconfig.json"),
      path.resolve(__dirname, "./packages/core/tests/tsconfig.json"),
      path.resolve(__dirname, "./packages/class-validator/tsconfig.json"),
      path.resolve(__dirname, "./packages/class-validator/tests/tsconfig.json"),
    ],
```

I suffer from this issue since v2.0 and its breaking change:
> by default we will now throw when a file is not in the project provided

**Comment by uniqueiniquity** [comment]
> Hmm, I'm not sure what the typescript compiler actually does to handle project references. I haven't actually used them yet (I've been meaning to migrate this monorepo to them, but I keep forgetting).
> 
> Perhaps @uniqueiniquity has more of an idea than I do.
> Is TS supposed to "include" all of the files in the referenced projects, or are they consumed via a different mechanism?

They are consumed by a different mechanism - in particular, the goal is that when project `a` references project `b`, project `a` does not include the .ts files in project `b`.

**Comment by bradzacher** [comment]
@19majkel94 - there is #806 which will make this easier by allowing you to supply a glob matcher for tsconfig files.

---

> They are consumed by a different mechanism - in particular, the goal of project references is that when project a references project b, project a does not include the .ts files in project b.

Interesting. That's good to know, thanks.

We should probably document this clearly so people don't assume they only need to provide one of the tsconfigs in the project.

**Comment by justnewbee** [comment]
facing the same issue here, it worked fine a while ago, and any project configuration files like webpack config was correctly parsed.

my project is like a main project plus a mono repo packages managed by lerna. the main project has it own src folder right under root, those files are OK.

all other files fail parsing.

here is a scratch of my project structure:

```
root/
  ├── packages/ <-- no good
  ├── src/ <-- good
  ├── webpack/ <-- no good
  ├── lerna.json
  ├── package.json
  └── tsconfig.json
```

and i use the latest versions of typescript and the linter

```
"@typescript-eslint/parser": "^2.0.0",
"typescript": "^3.5.3",
```

**Comment by bradzacher** [comment]
@justnewbee, did you try change suggested above?
What about the changes suggested in the [2.0.0 release notes](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.0.0)?

**Comment by lidianhao123** [comment]
@milesj I meet the same error. after add config to tsconfig.json is solved.
```json
{
    "allowJs": true,
}
```

**Comment by milesj** [comment]
Can confirm using an array of `tsconfig.json` fixed the problem. I just did it programmatically with this.

```ts
const pkg = getPackage();
const project = [fromRoot('tsconfig.json')];

if (pkg.workspaces) {
  project.push(
    ...glob.sync(pkg.workspaces.map(ws => path.join(ws, 'tsconfig.json')), {
      absolute: true,
      onlyFiles: true,
    }),
  );
}
```

**Comment by justnewbee** [comment]
> @justnewbee, did you try change suggested above?
> What about the changes suggested in the [2.0.0 release notes](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.0.0)?

tried, with no luck :(

**Comment by justnewbee** [comment]
sorry, i have to rollback to 1.13.0...

```
"@typescript-eslint/eslint-plugin": "^1.13.0",
"@typescript-eslint/parser": "^1.13.0",
```

i can do with the slowness, but i cannot do with failure...

hope there is a handy small little switch for us to turn the hard-fail off

**Comment by bradzacher** [comment]
Either your config is incorrect, or there is a bug.

Please open a new issue with detailed information (your tsconfig and eslint config) so we can look into helping you separately.

**Comment by bradzacher** [comment]
FYI #806 (glob selectors for `parserOptions.project`) has now been merged to master - so it'll be available on the `canary` tag on NPM imminently.

**Comment by milesj** [comment]
👏

**Comment by wintercounter** [comment]
I'm having the same issue. I have `include` field in my `tsconfig.json` as follows:
```
"include": [
    "C:\\Work\\Repos\\my-repo\\src\\**\\*"
  ],
```

Then I get the error:
```
0:0  error  Parsing error: If "parserOptions.project" has been set for @typescript-eslint/parser, C:\Work\Repos\my-repo\src\screens\Home\Sections\Top.tsx must be included in at least one of the projects provided.
```

I guess this is false and the pattern for some reason isn't matching (while in TS it does).

**Comment by bradzacher** [comment]
Please follow the steps in #890.

Locking this issue to direct all further conversation to that issue
